runners/flink/src/main/java/org/apache/beam/runners/flink/execution/SingletonSdkHarnessManager.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.flink.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.apache.beam.model.fnexecution.v1.ProvisionApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.ServerFactory;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactSource;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClientControlService;
+import org.apache.beam.runners.fnexecution.environment.ContainerManager;
+
+import javax.annotation.Nullable;
+
+/**
+ * A long-lived class to manage SDK harness container instances on behalf of
+ * shorter-lived Flink operators.  It is responsible for initializing and storing
+ * job resources for retrieval.
+ */
+public class SingletonSdkHarnessManager implements  SdkHarnessManager {
+  private static SingletonSdkHarnessManager ourInstance = new SingletonSdkHarnessManager();
+
+  /** Get the singleton instance of the harness manager. */
+  public static SingletonSdkHarnessManager getInstance() {
+    return ourInstance;
+  }
+
+  private final ServerFactory serverFactory;
+  private final ExecutorService executorService;
+  private final JobResourceManagerFactory jobResourceManagerFactory;
+
+  private SingletonSdkHarnessManager() {
+    this(
+        ServerFactory.createDefault(),
+        Executors.newCachedThreadPool(),
```

Comment:
Please verify that this ExecutorService does not prevent normal JVM termination.

If this is not the case, it's best to use `ForkJoinPool.commonPool()`.

Reply:
It sounds like [`newCachedThreadPool` uses `Executors.defaultThreadPool()`](https://stackoverflow.com/a/20057584), which itself creates non-daemon threads, though this is [not officially documented](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool--).

I've tested it on my macbook and it appears to work. If it works for you as well, let's assume this is safe.

Reply:
You can also use a custom `ThreadFactory` to make this explicit, as in `runners/direct/.../ExecutorServiceParallelExecutor` or `S3FileSystem` among others

Reply:
By the way, I had this backward and we specifically _do_ want new threads to be daemons. I had failed to submit any work to the thread pool in my local testing and never actually created any threads.

You'll either need to use a custom `ThreadFactory` or something like the common pool.